_posts/security-and-soa.md

```diff
@@ -0,0 +1,88 @@
+
+If the system uses Roles, we need to validate them whenever users attempt any operation: A user tries to do something, the system validates that they belong to a specific Role allowing that request.
+
+Similarly to tokens, roles might be self-contained in the token itself. For example, [.NET principals](https://learn.microsoft.com/en-us/dotnet/standard/security/principal-and-identity-objects), to be more precise, the Claim Principal, encapsulates roles in the form of claims. If the token is valid, another component can be responsible for extracting claims and presenting them to services for further usage.
+
+That poses some concerns. It creates some form of coupling between the business service that requires authorization and the security component that extracts claims from the token. There is little we can do about it: claims extraction is orthogonal to business services. It's a layer on top of every service. It's similar to logging. Due to security requirements, it presents itself as a mandatory infrastructure we must cope with.
+
+More coupling might surface if claims must be consistent. If we limit ourselves to extracting claims from the token, it might happen that someone else assigned a different set of roles to the user by the time a service checks them. If the business demands consistency, our only option becomes introducing some temporal coupling. The requirement forces us to refresh claims, thus assigned Roles or Groups, at every user request to guarantee that the system always uses the most up-to-date information.
+
+## Is that an issue?
+
+You bet it is. We now have a single point of failure. But again, it's primarily a business-caused issue. The business requirements force us in that direction. There is only so much we can do about it other than focusing on scaling out the active components and doing our best to provide high availability. In the end, it's what the business wants. A requirement such as the above states that the business favors a non-functioning system over a potentially insecure one.
+
+## It's not all, folks: tokens expire!
+
+So far, it went without saying that authentication and authorization happen at the system's outer boundaries. For example, web application front-ends validate authentication tokens and verify Roles. That implies a message sent on a queue by a front-end to a back-end service flies into the system unauthenticated.
+
+That is primarily because security tokens expire (for good reasons). For example, the Kerberos token lifetime is five minutes (IIRC). Active directory users need a new, freshly generated token every five minutes.
+
+Token expiration works in favor and against us at the same time. A token lifetime is a few minutes or more, which makes it behave like a cache. It helps in reducing the single point of failure issue highlighted above. If requirements allow, we can rely on the token "cache" to avoid going back to authentication and authorization components every time.
```

Comment:
```suggestion
Token expiration works for and against us at the same time. A token lifetime is a few minutes or more, which makes it behave like a cache. It helps in reducing the single point of failure issue highlighted above. If requirements allow, we can rely on the token "cache" to avoid going back to authentication and authorization components every time.
```